### PR TITLE
Hardshrink for Sparse Tensors

### LINKED ISSUE #16956: Hardshrink for Sparse Tensors
## 🚀 Feature

RuntimeError: hardshrink is not implemented for type torch.sparse.FloatTensor

## Motivation

We are experimenting with sparse neural networks. We want to model the connection between neurons with sparse matrices. At times, if the weights are too small we want to prune connections, i.e. set the element in the sparse matrix to zero via a hardshrink.

### PR DESCRIPTION
**Pull request by aocsa**
This PR implements hardshrink activation function on sparse tensors using the COO format. 

Resolves [#16956](https://github.com/pytorch/pytorch/issues/16956) . 

- [x] Hardshrink 
  - [x] CPU implementation
  - [x] CUDA implementation
  - [x] Unit tests for hardshrink
  - [x] Update documentation

### CODE REVIEW

**Comment by mattip** [comment]
The documentation task could be finished by adding something to `torch/_torch_docs.py`.

File: test/test_sparse.py [hunk 1]
```diff
@@ -2363,6 +2364,70 @@ def test_sparse_to_numpy(self):
...
+    @skipIfRocm
+    def test_hardshrink(self):
+
+        def test_hardshrink_edge_cases(device, dtype):
+            def h(values, l_expected):
+                for l, expected in l_expected.items():
+                    values_tensor = torch.tensor([float(v) for v in values], dtype=dtype, device=device).to_sparse()
+                    expected_tensor = torch.tensor([float(v) for v in expected], dtype=dtype, device=device)
+                    self.assertEqual(expected_tensor == torch.sparse.hardshrink(values_tensor, l).to_dense(),
+                                     torch.ones_like(values_tensor.to_dense(), dtype=torch.bool))
```
Comment (mattip) [inline test/test_sparse.py:2376]: I am not well-versed in sparse testing. Is there a way to check that the indices are indeed shrunk properly in addition to testing that the dense Tensor is as expected?

File: aten/src/ATen/native/sparse/Hardshrink.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "sparse_hardshrink_cpu", [&] {
+    auto lambd_val = lambd.to<scalar_t>();
+    cpu_kernel_vec(
+        iter,
+        [=](scalar_t self_val) {
+          return (self_val >= -lambd_val && self_val <= lambd_val) ? scalar_t(0)
+                                                                   : self_val;
+        },
+        [=](Vec256<scalar_t> self_val) {
+          return ((self_val < -lambd_val) | (self_val > lambd_val)) & self_val;
```
Comment (mattip) [inline aten/src/ATen/native/sparse/Hardshrink.cpp:25]: If I understand sparse iterators correctly, the second condition shrinks the indices, correct?

**Review by mattip — COMMENTED** [review]
A few general comments and questions, since I was requested to review. If the input is dense, will this behave as a clip function or will it fail?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -2390,6 +2390,17 @@
     CPU: gelu_backward_cpu
     CUDA: gelu_backward_cuda
 
+- func: _sparse_hardshrink(Tensor self, Scalar lambd=0.5) -> Tensor
```
[the remark this answers is not in the document]
Comment (aocsa) [reply aten/src/ATen/native/native_functions.yaml:2393]: Good catch! And yes, it is. I've just tried to match the current[ `torch.nn.Hardshink` API.](https://pytorch.org/docs/master/generated/torch.nn.Hardshrink.html), but, the tensorflow API looks more generic. Indeed, that is a good nice to have.

File: test/test_sparse.py (hunk 1 above)
Comment (aocsa) [reply test/test_sparse.py:2376]: Sure. Actually looking in detail my new task related to cuda sparse softmax, I figured out that I need to test this function with sparse tensors with a mix of sparse dims. I will do this in the next update.

File: aten/src/ATen/native/sparse/Hardshrink.cpp (hunk 2 above)
Comment (aocsa) [reply aten/src/ATen/native/sparse/Hardshrink.cpp:25]: Yes, the pytorch tensor iterators for CPU has two lambdas. Once for simple scalar processing and the other for vector processing. Both shrinks the values depending on the `lambd` input.

**Comment by aocsa** [comment]
> The documentation task could be finished by adding something to `torch/_torch_docs.py`.

Thanks @mattip for the review.  I will update the documentation and the PR to cover edges cases with sparse tensors with a mix of sparse dims.

**Comment by aocsa** [comment]
> A few general comments and questions, since I was requested to review. If the input is dense, will this behave as a clip function or will it fail?
I will fail. There is an assert for the input: `TORCH_INTERNAL_ASSERT(input_.is_sparse());`

**Comment by aocsa** [comment]
This PR is ready for review, the PR implements `Hardshrink` for CPU/CUDA over sparse tensors. cc @pearu

**Review by pearu — CHANGES_REQUESTED** [review]
This method of sparse hardshrink where the result values will contain zeros, does not address the concern raised in the original issue: the zero values from hardshrink remain in the sparse tensor values, see https://github.com/pytorch/pytorch/issues/16956#issuecomment-462382521 . I think this concern is valid and should be addressed.

There seems to be two approaches:
1. introduce some sparse tensor method that filters out the sparse tensor entries that match the following criteria: `<entry value> == <fill value>` so that users can apply this method to any sparse tensor depending on users need.
2. implement the fill value removal (including removal of the corresponding indices) within sparse_hardshrink.

Approach 2 has the potential to be more efficient than approach 1 but approach 1 would have wider usage for shrinking sparse tensors to a minimal size as well as for creating sparse tensors from dense tensors for given fill value.


In either way, the sparse hardshrink should use the following (pseudo)-algoritm:
```
m = at::native::hardshrink(input_values, lambda).nonzero()
result_indices = indices[m]
result_values = values[m]
```
but the implementation ought to avoid the creation of temporary tensor `m` if possible.

**Comment by ezyang** [comment]
merge looks messed up

**Pull request closed without merging**